doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,510 @@
+ * @param[in] mode Bulk operation error handling mode
+ * @param[out] object_statuses List of status for every object. Caller needs to
+ * allocate the buffer
+ *
+ * @return #SAI_STATUS_SUCCESS on success when all objects are removed or
+ * #SAI_STATUS_FAILURE when any of the objects fails to remove. When there is
+ * failure, Caller is expected to go through the list of returned statuses to
+ * find out which fails and which succeeds.
+ */
+typedef sai_status_t (*sai_bulk_get_inseg_entry_attribute_fn)(
+        _In_ uint32_t object_count,
+        _In_ const sai_inseg_entry_t *inseg_entry,
+        _In_ const uint32_t *attr_count,
+        _Inout_ sai_attribute_t **attr_list,
+        _In_ sai_bulk_op_error_mode_t mode,
+        _Out_ sai_status_t *object_statuses);
+```
+The existing sai_mpls_api_t structure has been expanded to accommodate the new MPLS bulking APIs:
+```
+/**
+ * @brief MPLS methods table retrieved with sai_api_query()
+ */
+typedef struct _sai_mpls_api_t
+{
+     sai_create_inseg_entry_fn                      create_inseg_entry;
+     sai_remove_inseg_entry_fn                      remove_inseg_entry;
+     sai_set_inseg_entry_attribute_fn               set_inseg_entry_attribute;
+     sai_get_inseg_entry_attribute_fn               get_inseg_entry_attribute;
+
++    sai_bulk_create_inseg_entry_fn                 create_inseg_entries;
++    sai_bulk_remove_inseg_entry_fn                 remove_inseg_entries;
++    sai_bulk_set_inseg_entry_attribute_fn          set_inseg_entries_attribute;
++    sai_bulk_get_inseg_entry_attribute_fn          get_inseg_entries_attribute;
+
+} sai_mpls_api_t;
+```
+
+#### Next Hop
+Full details about SAI Next Hop API and attributes are described here:
+https://github.com/opencomputeproject/SAI/blob/master/inc/sainexthop.h
+
+No modifications were made to the current SAI Next Hop API definition.
+
+The following existing attribute is now introduced to SONiC orchagent to facilitate MPLS NextHop functionality:
+```
+    /**
+     * @brief Push label
+     *
+     * @type sai_u32_list_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @condition SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_MPLS
+     */
+    SAI_NEXT_HOP_ATTR_LABELSTACK,
+```
+
+### Configuration and management
+This section should have sub-sections for all types of configuration and management related design. Example sub-sections for "CLI" and "Config DB" are given below. Sub-sections related to data models (YANG, REST, gNMI, etc.,) should be added as required.
+
+#### CLI Enhancements
+A new SONiC CLI command is introduced to configure interfaces for MPLS.
+
+    # Enable/disable MPLS per INTERFACE.
+    config interface mpls add|remove <intf-name>
+
+#### Config DB Enhancements
+
+##### INTERFACE
+The existing INTERFACE table is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
```

Comment:
as discussed in the yang subgroup meeting, all config db changes require to submit sonic-yang-mode under buildimage/src/sonic-yang-model. hld will not be approved until the yang model change is approved.

Reply:
We will look into the change in yang model.